compiler/src/iree/compiler/Dialect/HAL/Transforms/ResolveTopologyQueries.cpp

```diff
@@ -0,0 +1,228 @@
+                          << " has transparent access to all devices\n");
+  return true;
+}
+
+static bool tryAddSharedUsageBits(IREE::HAL::DeviceTopologyAttr topology,
+                                  IREE::HAL::DeviceOptimalAttr optimalAttr,
+                                  IREE::HAL::BufferUsageBitfield &bufferUsage,
+                                  IREE::HAL::MemoryTypeBitfield &memoryTypes,
+                                  DeviceAnalysis &deviceAnalysis,
+                                  Operation *fromOp) {
+  // If all affinities refer to the same device, we don't need to add any
+  // extra usage bits.
+  if (allReferToSameDevice(optimalAttr, deviceAnalysis, fromOp)) {
+    return true;
+  }
+  // Check if any device has transparent access to all other devices.
+  for (auto affinity : optimalAttr.getAffinities()) {
+    if (hasTransparentAccessToAll(topology, affinity, optimalAttr)) {
+      bufferUsage = bufferUsage | IREE::HAL::BufferUsageBitfield::Mapping;
+      return true;
+    }
+  }
+  return false;
+}
+
+static LogicalResult
+resolveMemoryPropertiesOp(AllocatorResolveMemoryPropertiesOp op,
+                          DeviceAnalysis &deviceAnalysis) {
+  LLVM_DEBUG(llvm::dbgs() << "[resolve_topology_queries] Op: " << op << "\n");
+
+  OpBuilder builder(op);
+  auto loc = op.getLoc();
+
+  // Get the default memory types and buffer usage based on lifetime.
+  auto memoryTypes = IREE::HAL::MemoryTypeBitfield::None;
+  auto bufferUsage = IREE::HAL::BufferUsageBitfield::None;
+  if (failed(deriveAllowedResourceBufferBits(loc, op.getLifetime(), memoryTypes,
+                                             bufferUsage))) {
+    return failure();
+  }
+
+  auto optimalAttr = dyn_cast_if_present<IREE::HAL::DeviceOptimalAttr>(
+      op.getAffinity().value_or(nullptr));
+  if (!optimalAttr) {
+    // If we don't have an optimal attribute, we just set the default
+    // memory types and buffer usage.
+    auto memoryTypeOp =
+        builder.create<IREE::HAL::MemoryTypeOp>(loc, memoryTypes);
+    auto bufferUsageOp =
+        builder.create<IREE::HAL::BufferUsageOp>(loc, bufferUsage);
+    op.replaceAllUsesWith(ValueRange{memoryTypeOp, bufferUsageOp});
+    op.erase();
+    return success();
+  }
+
+  // Get the module to access the topology attribute.
+  auto moduleOp = op->getParentOfType<mlir::ModuleOp>();
+  if (!moduleOp) {
+    return failure();
+  }
+
+  auto topologyAttr =
+      moduleOp->getAttrOfType<IREE::HAL::DeviceTopologyAttr>("stream.topology");
+
+  LLVM_DEBUG(llvm::dbgs() << "  -> Topology attr: " << topologyAttr << "\n");
+
+  // Try to resolve shared usage bits if possible.
+  if (!tryAddSharedUsageBits(topologyAttr, optimalAttr, bufferUsage,
+                             memoryTypes, deviceAnalysis, op.getOperation())) {
+    LLVM_DEBUG(llvm::dbgs() << "  -> Failed to add shared usage bits\n");
+    return failure();
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "  -> Successfully resolved memory properties "
```

Comment:
nit: for consistency with your other outputs above (and convention in LLVM/MLIR) use lowercase
e.g. above too ("-> failed to add", "-> topology attr")

(it may seem picky, but when scanning 100k line dumps the consistency really helps with search/scanning/human pattern recognition - differences in styling across passes or within a pass adds cognitive load)

```suggestion
  LLVM_DEBUG(llvm::dbgs() << "  -> successfully resolved memory properties "
```